libafl/src/events/launcher.rs

```diff
@@ -362,3 +365,240 @@ where
+    CF: FnOnce(
+        Option<S>,
+        CentralizedEventManager<LlmpRestartingEventManager<S, SP>, SP>,
+        CoreId,
+    ) -> Result<(), Error>,
+    S::Input: 'a,
+    MT: Monitor,
+    SP: ShMemProvider + 'static,
+    S: DeserializeOwned + UsesInput + 'a,
+{
+    /// The ShmemProvider to use
+    shmem_provider: SP,
+    /// The monitor instance to use
+    monitor: MT,
+    /// The configuration
+    configuration: EventConfig,
+    /// The 'main' function to run for each client forked. This probably shouldn't return
+    #[builder(default, setter(strip_option))]
+    run_client: Option<CF>,
+    /// The broker port to use (or to attach to, in case [`Self::spawn_broker`] is `false`)
+    #[builder(default = 1337_u16)]
+    broker_port: u16,
+    /// The list of cores to run on
+    cores: &'a Cores,
+    /// A file name to write all client output to
+    #[builder(default = None)]
+    stdout_file: Option<&'a str>,
+    /// A file name to write all client stderr output to. If not specified, output is sent to
+    /// `stdout_file`.
+    #[builder(default = None)]
+    stderr_file: Option<&'a str>,
+    /// The `ip:port` address of another broker to connect our new broker to for multi-machine
+    /// clusters.
+    #[builder(default = None)]
+    remote_broker_addr: Option<SocketAddr>,
+    /// If this launcher should spawn a new `broker` on `[Self::broker_port]` (default).
+    /// The reason you may not want this is, if you already have a [`Launcher`]
+    /// with a different configuration (for the same target) running on this machine.
+    /// Then, clients launched by this [`Launcher`] can connect to the original `broker`.
+    #[builder(default = true)]
+    spawn_broker: bool,
+    /// Tell the manager to serialize or not the state on restart
+    #[builder(default = true)]
+    serialize_state: bool,
+    #[builder(setter(skip), default = PhantomData)]
+    phantom_data: PhantomData<(&'a S, &'a SP)>,
+}
+
+#[cfg(all(unix, feature = "std", feature = "fork"))]
+impl<CF, MT, S, SP> Debug for CentralizedLauncher<'_, CF, MT, S, SP>
+where
+    CF: FnOnce(
+        Option<S>,
+        CentralizedEventManager<LlmpRestartingEventManager<S, SP>, SP>,
+        CoreId,
+    ) -> Result<(), Error>,
+    MT: Monitor + Clone,
+    SP: ShMemProvider + 'static,
+    S: DeserializeOwned + UsesInput,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Launcher")
+            .field("configuration", &self.configuration)
+            .field("broker_port", &self.broker_port)
+            .field("core", &self.cores)
+            .field("spawn_broker", &self.spawn_broker)
+            .field("remote_broker_addr", &self.remote_broker_addr)
+            .field("stdout_file", &self.stdout_file)
+            .field("stderr_file", &self.stderr_file)
+            .finish_non_exhaustive()
+    }
+}
+
+#[cfg(all(unix, feature = "std", feature = "fork"))]
+impl<'a, CF, MT, S, SP> CentralizedLauncher<'a, CF, MT, S, SP>
```

Comment:
Would be nice if we had some way to reuse the `Launcher` code instead of copy&paste? Maybe just a `EventManager` enum?

Reply:
yes when it will work i will unify the code, now it is easier to modify without changing the existing code

Reply:
Ok checked, this is impossible due to the callback trait bound, at compile time it cannot be both `CF: FnOnce(
        Option<S>,
        CentralizedEventManager<LlmpRestartingEventManager<S, SP>, SP>,
        CoreId,
    ) -> Result<(), Error>,` for centralized and `CF: FnOnce(
        Option<S>,
        LlmpRestartingEventManager<S, SP>,
        CoreId,
    ) -> Result<(), Error>,` for normal

Reply:
Maybe it can be an Enum with both types?